### PR TITLE
Loadable rewrite

### PR DESCRIPTION
**Pull request by tybug**
Mostly a code quality PR. Formalizes the relationship between loadables a bit more, renames `InfoLoadable` to `LoadableContainer` (leaving `InfoLoadable` as deprecated), adds equality methods and equality testing, adds abstract methods in places that make sense. Users shouldn't feel any change from this.

Except for maybe one thing, which is that `num_replays` was removed. Since loadables now implement `__iter__` in sensible places, users can just call `len(loadable_container)` or `len(loadable_container.all_replays())`.

Just found a good changelog in one of my commits (thanks past me):

* InfoLoadable renamed to LoadableContainer
* cache passed as part of super call
* `__eq__` required for Loadable
* `num_replays` method removed (should use `len(loadable_container.all_replays()` instead)
* all_replays required by LoadableContainer, not Loadable
* default implementation of `__iter__` and `__getitem__` in LoadableContainer
* ~~`LoadableContainer` implements new default `__contains__` method~~ wasn't comfortable with the correctness of the implementation, leaving this for a later date
* `check.all_replays()` now returns the expected value (`loadables1 + loadables2`), new method `#all_replays1` which returns only the replays in `loadables1`.

### CODE REVIEW

File: tests/test.py [hunk 1]
```diff
@@ -266,6 +266,40 @@ def test_incorrect_key(self):
...
+        cls.cg.load_info(cls.map2)
+
+    def test_identical_equality(self):
+        self.assertEqual(self.user, self.user1)
+        self.assertNotEqual(self.user, self.user2)
+        self.assertNotEqual(self.user, self.user3)
+
+        self.assertEqual(self.map, self.map1)
+        self.assertNotEqual(self.map, self.map2)
+        self.assertNotEqual(self.map, self.map3)
```
Comment (InvisibleSymbol) [inline tests/test.py]: I would like to have them split in separate tests
```python
def test_user_equality(self):
    self.assertEqual(self.user, self.user1)

def test_user_inequality1(self):
    self.assertNotEqual(self.user, self.user2)

def test_user_inequality2(self):
    self.assertNotEqual(self.user, self.user3)
```
bit inefficient but having them separate tests is useful since it instantly shows what is broken, not only that something is broken.

File: circleguard/loadable.py [hunk 2]
```diff
@@ -659,6 +622,34 @@ def load(self, loader, cache):
...
+        the same replay".
+
+        TODO possible false positive if a user overwrites their score inbetween
+        loading two otherwise identical replay maps. Similar situation to
+        ReplayPath equality. Could equality check replay data instead if both
+        are loaded.
+        """
+        if not isinstance(loadable, ReplayMap):
+            return False
+        return self.map_id == loadable.map_id and self.user_id == loadable.user_id and self.mods == loadable.mods
```
[the remark this answers is not in the document]
Comment (tybug) [reply circleguard/loadable.py]: that only works when the replay is loaded, returning `True` for all unloaded replays

File: circleguard/loadable.py [hunk 3]
```diff
@@ -45,38 +45,58 @@ def load(self, loader, cache):
...
+    def all_replays(self):
+        pass
+
+    # def __contains__(self, loadable):
+    #     our_replays = self.all_replays()
+    #     if isinstance(loadable, LoadableContainer):
+    #         # if we contain all the Replays in the loadable container, we contain
+    #         # the loadable container as well
+    #         return all([replay in our_replays for replay in loadable.all_replays()])
+    #     return loadable in our_replays
```
[the remark this answers is not in the document]
Comment (tybug) [reply circleguard/loadable.py]: wasn't really thinking when I did this (well I was but didn't feel like making a new branch lol), I agree

File: circleguard/loadable.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (InvisibleSymbol) [reply circleguard/loadable.py]: oh true, guess this works then.

File: tests/test.py (hunk 1 above)
Comment (tybug) [reply tests/test.py]: That's a bit too wordy for me, and not really how the rest of the tests are set up. I think they should be split into user and map, but nothing beyond that

File: tests/test.py [hunk 4]
```diff
@@ -266,6 +266,40 @@ def test_incorrect_key(self):
...
+        cls.user1 = User(2757689, num=2, cache=False)
+        cls.user2 = User(2757689, num=2, mods=Mod.HT)
+        cls.user3 = User(2757689, num=1)
+
+        cls.map = Map(1754777, num=4) #sidetracked day: umbre, karthy, -duckleader-, toy
+        cls.map1 = Map(1754777, num=4, cache=False)
+        cls.map2 = Map(1754777, num=4, mods=Mod.HD)
+        cls.map3 = Map(1754777, num=2)
+
+        cls.r = ReplayMap(1754777, 2766034) # umbre +HDHR on sidetracked day
```
[the remark this answers is not in the document]
Comment (tybug) [reply tests/test.py]: was going to use it for #contains

File: tests/test.py [hunk 5]
```diff
@@ -266,6 +266,39 @@ def test_incorrect_key(self):
...
+    def setUpClass(cls):
+        cls.cg = Circleguard(KEY)
+
+        cls.user = User(2757689, num=2) # toy, #1=sidetracked day, #2=View of The River Styx
+        cls.user1 = User(2757689, num=2, cache=False)
+        cls.user2 = User(2757689, num=2, mods=Mod.HT)
+        cls.user3 = User(2757689, num=1)
+
+        cls.map = Map(1754777, num=4) #sidetracked day: umbre, karthy, -duckleader-, toy
+        cls.map1 = Map(1754777, num=4, cache=False)
```
Comment (InvisibleSymbol) [inline tests/test.py]: why is this `cache=False`?
